### PR TITLE
Bump compact_index gem to 0.13.0

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
I'd like to get the `compact_index` gem updated to pickup the fix in https://github.com/rubygems/compact_index/pull/26.

So as a first step I'm opening this draft to make sure tests pass.

### CODE REVIEW

**Comment by sonalkr132** [comment]
sort any space trim got us :sweat_smile: will you be fixing tests?

I was not able to reproduce info checksum mismatch last week.. I will give this another day. We will move ahead with the versions file update if I can't reproduce it.

**Comment by deivid-rodriguez** [comment]
I'm guessing it's the extra whitespace cleanup that was added to the patch: https://github.com/rubygems/compact_index/pull/26/files#diff-814b0d2c526716c372b5bdc6ce59cf28R50.

Should this be fixed here or that change reverted in the gem?

**Comment by sonalkr132** [comment]
We need to update these failing tests. No good reason to keep spaces in the info response.

**Comment by deivid-rodriguez** [comment]
Ok!

File: test/unit/update_versions_file_test.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ class UpdateVersionsFileTest < ActiveSupport::TestCase
       end
 
       should "include platform release" do
-        expected_output = "rubyrubyruby 0.0.1-jruby,0.0.1 qw212r\n"
+        expected_output = "rubyrubyruby 0.0.1,0.0.1-jruby qw212r\n"
```
Comment (deivid-rodriguez) [inline test/unit/update_versions_file_test.rb:47]: There's also this "ordering change" that I'm not sure where it comes from :man_shrugging:

Comment (simi) [reply test/unit/update_versions_file_test.rb:47]: Maybe related to https://github.com/rubygems/compact_index/commit/63187755018e11ffc22b6e56d8463d13898b1a0e?

Comment (sonalkr132) [reply test/unit/update_versions_file_test.rb:47]: yes, simi is correct. The order doesn't matter at the moment, but allow me to explain this to ensure all of us are on the same page.
You may already know that bundler uses the checksum of the local gem info file and checksum reported by the versions file to decide if it should make a request to rg.org to update it's gem info file.
The info checksum reported in the versions endpoint for the gem will change when a gem releases or yanks a version. info checksum as reported by versions endpoint always has to be one calculated for the latest update (created_at/id column of versions table), so that the client can see checksum mismatch and fetches the latest changes.
Before [this change](https://github.com/rubygems/rubygems.org/pull/1454/files), we were sorting versions by [number, platform], and it meant we were using the checksum of the last version by number and platform (not last created at) making the client miss info updates if someone release `0.1.0` after `2.0.1`. 

Above linked PR would the second instance of us trying to fix something broken in `compact_index` by overriding the behavior in rubygems.org. Most plausible reason for doing this seems to be avoiding the hassle of releasing a new version. 
It makes me question if compact_index really has to be separate repo.

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: Thanks for explaining, makes sense.

Note that `bundler` also uses the `compact_index` gem for testing, I guess that's why it is in a separate repo.

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: So, as per the code comment in the commit you linked to, we could now get rid of https://github.com/rubygems/rubygems.org/pull/1454/files#diff-2cf34b93c3257e00201d866038d19583R91-R93, right?

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: Code has moved, actually, it's here now:

https://github.com/rubygems/rubygems.org/blob/995069b09bc138fe662ff78af9b5a368b7c64371/app/models/gem_info.rb#L64-L65

Comment (sonalkr132) [reply test/unit/update_versions_file_test.rb:47]: > bundler also uses the compact_index gem for testing

I would consider re-evaluating if [these need to use compact_index API](https://github.com/rubygems/bundler/blob/bf7df421ee6311fcc17e6dc46d204236f33b4b3f/spec/support/artifice/compact_index.rb#L94-L116) or we can just hardcode the response.

> we could now get rid of https://github.com/rubygems/rubygems.org/pull/1454/files#diff-2cf34b93c3257e00201d866038d19583R91-R93, right?

We could. I would lean towards not changing anything tho. Given the checksum mismatch we are seeing, we still seem to be figuring out all the edge cases.

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: Sure. I only found out recently that rubygems.org also used this gem, I thought it was just a little gem to mock rubygems.org responses. When I found about it, it felt nice that we were using the same thing but if you feel there's value in replacing it, I'm fine with re-evaluating it, sure.

Comment (sonalkr132) [reply test/unit/update_versions_file_test.rb:47]: rubygems.org is the only user if you ask me. I see `vcr_cassettes` in bundler repo, they could be used for mocking rubygems.org response. are you sure you are using the latest compact_index version?

Comment (sonalkr132) [reply test/unit/update_versions_file_test.rb:47]: > there's value in replacing it

value being not keeping a dependency you don't need, even if it is just a test/development dependency.

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: > are you sure you are using the latest compact_index version?

No, we are using 0.11.0, we haven't upgraded yet.

> value being not keeping a dependency you don't need, even if it is just a test/development dependency.

It depends on the alternative, removing a dependency is not good per se. But as I said, I'm opening to exploring alternatives :+1:

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: I just tried upgrading and our specs crash with the new version: https://github.com/rubygems/rubygems/pull/3615. Needs investigation.

Comment (sonalkr132) [reply test/unit/update_versions_file_test.rb:47]: > `NoMethodError: undefined method 'gsub' for #<Gem::Requirement:0x00005628ab6bfb28>`

from the following:
```
 def ruby_version_line
      join_multiple(ruby_version)
 end
```
AFAIU, compact_index only supports string as requirement. Removing compact_index from bundler is going to be harder than I assumed :sweat_smile: 
will look into removing above mentioned patch from rubygems.org. We should probably move more things to compact_index (like generating new versions file) both from rubygems.org and bundler if we do have to keep compact_index the gem around.

Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: That last issue was solved in bundler, and my original comment was explained too, so marking this as resolved :)

**Comment by deivid-rodriguez** [comment]
I believe this one should be ready now? After this is deployed, I'll go ahead and revert https://github.com/rubygems/bundler/pull/7554 before the next bundler release.

**Comment by sonalkr132** [comment]
> I believe this one should be ready now?

Kind of. versions endpoint will end up with two different formats, which technically shouldn't break anything but best to merge this when we are updating versions file.

I am having a busy, and not able to find time for info checksum mismatch at the moment. Checksum fix will most likely be delayed, it affects only a couple hundred versions. I was kinda hoping we could finish #2343 before update or else those 3548 versions will fall in the dynamic part of versions endpoint as well. Maybe @colby-swandale can take look if he can find time for it (#2343) :pray:

**Comment by indirect** [comment]
I am ok with shipping this and cleaning up later, assuming that the only bad effect of checksum mismatches is that Bundler downloads the /info/ for those gems when it doesn't actually need to. 👍

**Comment by sonalkr132** [comment]
Since info response will change from:
> 1.6.7.rc2-x64-mingw32 mini_portile:~> 0.7.0.rc2|checksum:3b03a61d7c22883673d5bf3794019f41184c1b0b4686d99ee2e122661e99f88f,ruby:< 2.3, >= 1.9.2,rubygems:> 1.3.1

to:
> 1.6.7.rc2-x64-mingw32 mini_portile:~> 0.7.0.rc2|checksum:3b03a61d7c22883673d5bf3794019f41184c1b0b4686d99ee2e122661e99f88f,ruby:< 2.3% >= 1.9.2,rubygems:> 1.3.1

we will need to update info_checksum of the latest (or all) version of these gems.

File: test/integration/api/compact_index_test.rb [hunk 2]
```diff
@@ -118,10 +118,10 @@ def etag(body)
...
     expected = <<~VERSIONS_FILE
       ---
-      1.0.0 |checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,rubygems:>= 2.6.3
-      2.0.0 gemA1:= 1.0.0|checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,rubygems:>= 2.6.3
-      1.2.0 |checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,ruby:>= 2.0.0,rubygems:>1.9
-      2.1.0 gemA1:= 1.0.0,gemA2:= 1.0.0|checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,ruby:>= 2.0.0,rubygems:>=2.0
+      1.0.0 |checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,rubygems:>=2.6.3
+      2.0.0 gemA1:=1.0.0|checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,rubygems:>=2.6.3
+      1.2.0 |checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,ruby:>=2.0.0,rubygems:>1.9
+      2.1.0 gemA1:=1.0.0,gemA2:=1.0.0|checksum:b5d4045c3f466fa91fe2cc6abe79232a1a57cdf104f7a26e716e0a1e2789df78,ruby:>=2.0.0,rubygems:>=2.0
```
Comment (sonalkr132) [inline test/integration/api/compact_index_test.rb:124]: I am not sure if we want to fix the redundant space issue here. This will mean we will have update info_checksum of every gem.

**Comment by indirect** [comment]
I don’t think it will hurt anything to update all the checksums, as long as past and current bundlers still parse the output successfully

**Comment by sonalkr132** [comment]
> it will hurt anything to update all the checksums

Yes, it won't break anything but updating all checksums makes this change bigger than this has to be (at least in my head). 
If we are updating all checksums, we need to decide if we want to overwrite info_checksum of all versions (something we did in the backfill the very first time) or just [last version as picked here](https://github.com/rubygems/rubygems.org/blob/master/app/models/gem_info.rb#L62). 
Neither of the options is ideal and at the same time, which option gets picked doesn't matter much as long as final info_checksum show in versions file matches the info response.

**Comment by deivid-rodriguez** [comment]
I'm not sure of what running a whole DB update like this would imply exactly,but conceptually it also feel bigger in my head.

Looking at https://github.com/rubygems/compact_index/pull/26, I believe the whitespace change was intentional, but the side effects of it (requiring an update of all checksums) was overlooked. It was also unrelated to the fix being proposed there. So my proposal would be to revert that specific change and release a new patch level version of compact_index so that we don't need to update the whole DB.

**Comment by sonalkr132** [comment]
>  the whitespace change was intentional, but the side effects of it (requiring an update of all checksums) was overlooked. 

whitespace removal was intentional but I overlooked the fact that dependency requirements (not rubygem/ruby req) will change because of it. I updated `deps_line` as it looked like a duplicate code at the time.

I will revert this for `dep_line` change in compact_index.

**Comment by sonalkr132** [comment]
@deivid-rodriguez do you think we should revert whitespace change from ruby/gem requirement as well for consistency (to match dependency requirement)?

**Comment by deivid-rodriguez** [comment]
I would only revert the `gsub` call for consistency yeah.

File: test/unit/update_versions_file_test.rb (hunk 1 above)
Comment (sonalkr132) [inline test/unit/update_versions_file_test.rb:47]: > There's also this "ordering change" that I'm not sure where it comes from man_shrugging

>  The order doesn't matter at the moment, 

I was wrong about this as well. The order does matter as the checksum of these two lines is different.
```
2.6.5 :014 > Digest::MD5.hexdigest("rubyrubyruby 0.0.1-jruby,0.0.1 qw212r\n")
 => "b3eb3b29e9b520cb64bb3d4038e06bbb" 
2.6.5 :015 > Digest::MD5.hexdigest("rubyrubyruby 0.0.1,0.0.1-jruby qw212r\n")
 => "c3892a889a4d9c7a2d2672f63d7e9a95" 
```

Comment (sonalkr132) [inline test/unit/update_versions_file_test.rb:47]: info response should not change because of this. `gem_lines` is only being used to generate versions file. @deivid-rodriguez please verify this anyway.

I use the following steps to verify checksum:
```
$ rake compact_index:update_versions_file
$ cat lib/versions_verify_checksum.rb
file = File.open("/home/aditya/rubygems.org/log/versions-updated.log", 'w')
File.readlines("/home/aditya/rubygems.org/config/versions.list").drop(2).each do |line|
  name, versions, info_checksum = line.split(" ")
  calc_info_checksum = Digest::MD5.hexdigest(CompactIndex.info(GemInfo.new(name).compact_index_info))
  if calc_info_checksum != info_checksum
    puts "#{name} #{info_checksum} #{calc_info_checksum}"
    file.write("#{name} #{info_checksum} #{calc_info_checksum}\n")
  end
end
file.close
$ rails runner lib/versions_verify_checksum.rb
```
Haven't done this so far for compact_index update.

File: test/integration/api/compact_index_test.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply test/integration/api/compact_index_test.rb]: Concern addressed as discussed by upgrading all the way to 0.13.0.

File: test/unit/update_versions_file_test.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: Ok, I'll double check it!

**Comment by sonalkr132** [comment]
I have verified that checksum mismatch count doesn't increase from [the existing total of 313](https://gist.github.com/sonalkr132/617bb282a61cc03a7c41ad6337e59159) after we have updated compact_index and run the rake task to update checksums. All :heavy_check_mark: from my side.

**Comment by deivid-rodriguez** [comment]
Nice! :heart: 

Should I still verify anything on my side?

**Comment by sonalkr132** [comment]
Yes, please run the rake task added in #2370 on this branch (deivid-rodriguez:bump_compact_index), update the versions file and make sure versions `where("required_ruby_version like '%,%' or required_rubygems_version like '%,%'")` required ruby and rubygems versions are still respected by the client and there is no info checksum mismatch.
checking a few versions taken randomly from ~5k total should be good enough.

**Comment by deivid-rodriguez** [comment]
Testing went good.

I applied the task to this PR, and run bundler against it with a few gems with multi ruby version or rubygems version requirements.

Initially I got the following error:

```
....
[Bundler::CompactIndexClient] skipping updating info for ice_nine since the versions checksum matches the local checksum
Looking up gems ["memoizable"]
[Bundler::CompactIndexClient] execution_mode=
[Bundler::CompactIndexClient] dependencies(["memoizable"])
[Bundler::CompactIndexClient] update_info(memoizable)
[Bundler::CompactIndexClient] skipping updating info for memoizable since the versions checksum matches the local checksum
There was an error parsing the metadata for the gem cfer (0.5.0.pre.rc1): Gem::Requirement::BadRequirementError
Illformed requirement [">= 2.2.5&~> 2.2"]
The metadata was [["checksum", ["ac1757b459afad42b4408dd2f37bd0d3e48857cbd758c2ea1debfce9b7b77ad3"]], ["ruby", [">= 2.2.5&~> 2.2"]], ["rubygems", ["> 1.3.1"]]]
```

This is expected and caused by https://github.com/rubygems/bundler/pull/7554. Reverting that PR gets everything working as expected, with proper behavior according to the requirements.

Also tried an older `bundler` version and it also behaves correctly.

Also verified that rubygems.org upgrading to the latest version of the compact index gem makes all releases of bundler understand multiple ruby & rubygems requirements and start considering them for resolution.

For example, using bundler 1.17.3 on ruby 2.7 against a `Gemfile` including only `gem "sorbet", "0.4.5481"` which is a sorbet version that has `>= 2.3.0, < 2.7.0.preview1` ruby requirements, I get the following results:

### Against current rubygems.org

Bundler incorrectly resolves to the specified version (because it disregards the `< 2.7.0.preview1` requirement, and fails at installation time

```
Fetching gem metadata from http://rubygems.org/..
Resolving dependencies...
Using bundler 2.2.0.dev
Fetching sorbet-static 0.4.4581 (x86_64-linux)
Installing sorbet-static 0.4.4581 (x86_64-linux)
Gem::RuntimeRequirementNotMetError: sorbet-static requires Ruby version >= 2.3.0, < 2.7.0.preview1. The current ruby version is 2.7.1.83.
An error occurred while installing sorbet-static (0.4.4581), and Bundler cannot continue.
Make sure that `gem install sorbet-static -v '0.4.4581' --source 'http://rubygems.org/'` succeeds before bundling.

In Gemfile:
  sorbet was resolved to 0.4.4581, which depends on
    sorbet-static
```

### Against a locally run rubygems.org using this PR

Resolution fails with a proper error message because the `< 2.7.0.preview1` requirement is considered:

```
$ bundle _1.17.3_
Fetching gem metadata from http://rubygems.org/..
Resolving dependencies...
Bundler could not find compatible versions for gem "ruby":
  In Gemfile:
    ruby

    sorbet (= 0.4.4581) was resolved to 0.4.4581, which depends on
      ruby (< 2.7.0.preview1, >= 2.3.0)

```

The error message is confusing because it doesn't say the conflicting ruby, just "ruby". That will be fixed on the latest bundler though:

```
$ bundle
Fetching gem metadata from http://rubygems.org/..
Resolving dependencies...
Bundler found conflicting requirements for the Ruby version:
  In Gemfile:
    Ruby (= 2.7.1.83)

    sorbet (= 0.4.4581) was resolved to 0.4.4581, which depends on
      Ruby (< 2.7.0.preview1, >= 2.3.0)
```

Finally, all the tests I had done so far, I did them without generating any `versions.list` file at all, so that it was generated "on the fly" everytime. That's why I had to add `bundle config timeout 60` to my commands :sweat_smile:.

So I run `bin/rake compact_index:update_versions_file` and made a couple more tests. Response was obviously much faster, but also correct. First time I run the command, all `info/` endpoints were requested because I had deleted the cache. Second time, without deleting the cache, no `info/` endpoints were requested because all checksums matched.

File: test/unit/update_versions_file_test.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply test/unit/update_versions_file_test.rb:47]: I'm not sure if I need to test anything else here, but the gems I used for testing have a lot of platform specific versions, and I didn't see any issues with missing checksums so I believe it's fine.

If I understand correctly, this is a change in the versions file formatting, but it doesn't change the `/info` endpoint, so it should affect checksums. Is that correct?

**Comment by deivid-rodriguez** [comment]
I reverted the fix for this in bundler!

**Comment by indirect** [comment]
awesome, thanks for testing this so carefully everyone!
